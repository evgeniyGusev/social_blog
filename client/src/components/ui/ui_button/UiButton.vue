<template>
  <button
    v-bind="$attrs"
    type="button"
    :class="[
      'ui-button',
      {
        'ui-button--small': size === 'small',
        'ui-button--medium': size === 'medium',
        'ui-button--large': size === 'large',
        'ui-button--primary': template === 'primary',
        'ui-button--secondary': template === 'secondary',
        'ui-button--tertiary': template === 'tertiary',
        'ui-button--link': template === 'link',
        'ui-button--danger': template === 'danger',
        'ui-button--disabled': disabled,
        'ui-button--rounded': rounded,
        'ui-button--is-loading': isLoading,
      },
    ]"
    :disabled="disabled"
  >
    <div class="ui-button-text">
      <slot></slot>
    </div>
  </button>
</template>

<script setup lang="ts">
import { IUiButtonProps } from './interface';

withDefaults(defineProps<IUiButtonProps>(), {
  size: 'medium',
  template: 'primary',
});
</script>

<style scoped lang="scss">
.ui-button {
  position: relative;
  overflow: hidden;
  display: inline-flex;
  flex-direction: row;
  align-items: center;
  justify-content: center;
  padding: 0.65rem 1rem;
  border: 1px solid transparent;
  border-radius: $default-radius;
  cursor: pointer;
  transition: all 0.2s ease-in-out;

  &:not(.ui-button--rounded):focus {
    outline: #000 solid 3px;
  }

  .ui-button-text {
    display: flex;
    align-items: center;
  }

  &--disabled {
    .ui-button-text {
      opacity: 0.3;
    }
  }

  &-spinner {
    position: absolute;
    top: 50%;
    left: 42%;
    transform: translate(-50%, -42%);
  }

  &--small {
    padding: 0.25rem 0.5rem;
  }

  &--large {
    padding: 1rem 1.5rem;
  }

  &--primary {
    background-color: $color-primary;

    .ui-button-text {
      color: #fff;
    }

    &:not(:disabled):hover {
      background-color: $color-light;

      .ui-button-text {
        color: $color-primary;
      }
    }

    &.ui-button--disabled {
      background-color: $color-light;
      cursor: not-allowed;
    }
  }

  &--secondary {
    background-color: #fff;
    border-color: $color-primary;

    &:not(:disabled):hover {
      background-color: $color-light;
      border-color: transparent;
    }

    &.ui-button--disabled {
      border-color: $color-light;
      cursor: not-allowed;
    }
  }

  &--tertiary {
    background-color: #fff;
    border-color: transparent;
    transition: background-color 0.2s ease-in-out;

    &:not(:disabled):hover {
      background-color: $color-light;
    }

    &.ui-button--disabled {
      cursor: not-allowed;
    }
  }

  &--link {
    padding: 0;
    background-color: transparent;
    border-color: transparent;
    text-decoration: underline;
    text-decoration-color: $color-primary;
    transition: text-decoration-color 0.2s ease-in-out;

    &:not(:disabled):hover {
      text-decoration-color: $color-light;
    }

    &.ui-button--disabled {
      text-decoration-color: $color-secondary;
      cursor: not-allowed;
    }
  }

  &--rounded {
    border-radius: 100%;
    padding: 0;

    .ui-button-spinner {
      left: 25%;
    }
  }

  &--is-loading {
    &::after {
      content: '';
      position: absolute;
      top: 50%;
      left: 50%;
      width: 24px;
      height: 24px;
      margin: -12px 0 0 -12px;
      border: 4px solid transparent;
      border-top: 4px solid $color-primary;
      border-radius: 50%;
      animation: spin 0.75s linear infinite;
      box-sizing: border-box;
      transform-origin: 50% 50%;
    }

    @keyframes spin {
      0% {
        transform: rotate(0deg);
      }
      100% {
        transform: rotate(360deg);
      }
    }
  }
}
</style>
