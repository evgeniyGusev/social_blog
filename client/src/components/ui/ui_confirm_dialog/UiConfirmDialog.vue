<template>
  <ui-dialog :model-value="isDialogVisible" class="ui-confirm-dialog" :closable="false">
    <slot name="header">
      {{ content.head }}
    </slot>

    <slot>
      {{ content.body }}
    </slot>

    <nav class="ui-confirm-dialog-buttons">
      <slot name="buttons">
        <ui-button :disabled="disabled" :loading="loading" size="large" type="button" @click="confirm">
          {{ content.confirm }}
        </ui-button>

        <ui-button template="secondary" size="large" type="button" @click="cancel">
          {{ content.cancel }}
        </ui-button>
      </slot>
    </nav>
  </ui-dialog>
</template>

<script setup lang="ts">
import { reactive, ref } from 'vue';

import { IUiConfirmDialogProps, IUiConfirmDialogContent } from '@/components/ui/ui_dialog/interfaces.ts';

defineProps<IUiConfirmDialogProps>();

const emit = defineEmits(['confirm', 'cancel', 'open']);

const DEFAULT_CONTENT: IUiConfirmDialogContent = {
  body: 'Вы действительно хотите выполнить это действие?',
  cancel: 'Нет',
  confirm: 'Да',
  head: 'Подтверждение действия',
};

const isDialogVisible = ref(false);

const content = reactive(DEFAULT_CONTENT);

const callbacks = reactive({ resolve: (value: any) => value, reject: () => {} });

function resetCallbacks(resolve: (value: any) => void = () => {}, reject: () => void = () => {}): void {
  callbacks.resolve = resolve;
  callbacks.reject = reject;
}

function applyIncomingContent(incomingContent: IUiConfirmDialogContent = DEFAULT_CONTENT): void {
  Object.keys(DEFAULT_CONTENT).forEach((key) => {
    content[key as keyof IUiConfirmDialogContent] = incomingContent[key as keyof IUiConfirmDialogContent];
  });
}

const open = (incomingContent: IUiConfirmDialogContent = DEFAULT_CONTENT) =>
  new Promise((resolve, reject) => {
    applyIncomingContent(incomingContent);

    resetCallbacks(resolve, reject);

    isDialogVisible.value = true;

    emit('open');
  });

const confirm = () => {
  callbacks.resolve?.(true);

  resetCallbacks();

  isDialogVisible.value = false;

  emit('confirm');
};

const cancel = () => {
  callbacks.reject?.();

  resetCallbacks();

  isDialogVisible.value = false;

  emit('cancel');
};

defineExpose({
  cancel,
  confirm,
  open,
});
</script>

<style scoped lang="scss">
.ui-confirm-dialog-buttons {
  display: flex;
  flex-direction: column;
  gap: 1rem;
}
</style>
