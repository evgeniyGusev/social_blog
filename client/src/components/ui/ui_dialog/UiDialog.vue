<template>
  <teleport to="body">
    <transition name="dialog">
      <div v-if="modelValue" class="dialog" @click="closeDialog($event)" @keyup.esc="$emit('update:modelValue', false)">
        <div class="dialog-content">
          <slot />
        </div>
      </div>
    </transition>
  </teleport>
</template>

<script setup lang="ts">
defineProps<{
  modelValue: boolean;
}>();

const emit = defineEmits(['update:modelValue']);

const closeDialog = (event: Event) => {
  if (event?.target?.className !== 'dialog') {
    return;
  }

  emit('update:modelValue', false);
};
</script>

<style>
:root {
  --dialog-content-max-width: 400px;
}
</style>

<style scoped lang="scss">
.dialog {
  z-index: 999;
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  display: flex;
  align-items: center;
  justify-content: center;
  background-color: rgba(0, 0, 0, 0.7);
  backdrop-filter: blur(0.3rem);

  &-content {
    min-width: 364px;
    max-width: var(--dialog-content-max-width);
    padding: 2rem;
    background-color: #fff;
    border-radius: $default-radius;
    box-shadow: 0 0 1rem rgba(0, 0, 0, 0.5);
  }
}
</style>
