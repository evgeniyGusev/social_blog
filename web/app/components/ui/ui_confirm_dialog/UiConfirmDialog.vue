<template>
  <ui-dialog :model-value="isDialogVisible" class="ui-confirm-dialog" :closable="false">
    <div data-test="ui-confirm-dialog-header" class="ui-confirm-dialog-header">
      <slot name="header">
        {{ content.head }}
      </slot>
    </div>

    <div class="ui-confirm-dialog-body">
      <slot>
        {{ content.body }}
      </slot>
    </div>

    <nav class="ui-confirm-dialog-buttons">
      <slot name="buttons">
        <ui-button :disabled="disabled" :loading="loading" type="button" @click="confirm">
          {{ content.confirm }}
        </ui-button>

        <ui-button template="secondary" type="button" @click="cancel">
          {{ content.cancel }}
        </ui-button>
      </slot>
    </nav>
  </ui-dialog>
</template>

<script setup lang="ts">
import { reactive, ref } from 'vue';

import { IUiConfirmDialogProps, IUiConfirmDialogContent } from '@/components/ui/ui_dialog/interfaces.ts';

defineProps<IUiConfirmDialogProps>();

const emit = defineEmits(['confirm', 'cancel', 'open']);

const DEFAULT_CONTENT: IUiConfirmDialogContent = {
  body: 'Вы действительно хотите выполнить это действие?',
  cancel: 'Нет',
  confirm: 'Да',
  head: 'Подтверждение действия',
};

const isDialogVisible = ref(false);

const content = reactive(DEFAULT_CONTENT);

const callbacks = reactive({ resolve: (value: any) => value, reject: () => {} });

function resetCallbacks(resolve: (value: any) => void = () => {}, reject: () => void = () => {}): void {
  callbacks.resolve = resolve;
  callbacks.reject = reject;
}

function applyIncomingContent(incomingContent: IUiConfirmDialogContent = DEFAULT_CONTENT): void {
  Object.keys(DEFAULT_CONTENT).forEach((key) => {
    content[key as keyof IUiConfirmDialogContent] = incomingContent[key as keyof IUiConfirmDialogContent];
  });
}

const open = (incomingContent: IUiConfirmDialogContent = DEFAULT_CONTENT) =>
  new Promise((resolve, reject) => {
    applyIncomingContent(incomingContent);

    resetCallbacks(resolve, reject);

    isDialogVisible.value = true;

    emit('open');
  });

const confirm = () => {
  callbacks.resolve?.(true);

  resetCallbacks();

  isDialogVisible.value = false;

  emit('confirm');
};

const cancel = () => {
  callbacks.reject?.();

  resetCallbacks();

  isDialogVisible.value = false;

  emit('cancel');
};

defineExpose({
  cancel,
  confirm,
  open,
});
</script>

<style scoped lang="scss">
.ui-confirm-dialog {
  &-header {
    margin-bottom: 1rem;
    font-size: 2rem;
    text-align: center;
    font-weight: 700;
  }

  &-body {
    margin-bottom: 1rem;
    font-size: 1.5rem;
    text-align: center;
  }

  &-buttons {
    display: flex;
    flex-direction: column;
    gap: 0.5rem;
  }
}
</style>
