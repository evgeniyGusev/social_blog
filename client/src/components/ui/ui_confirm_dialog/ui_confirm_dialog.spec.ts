import { describe, expect, it, beforeEach } from 'vitest';
import { shallowMount } from '@vue/test-utils';

import UiConfirmDialog from './UiConfirmDialog.vue';
import UiButton from '../ui_button/UiButton.vue';
import UiDialog from '../ui_dialog/UiDialog.vue';

let wrapper: any;

beforeEach(() => {
  wrapper = shallowMount(UiConfirmDialog, {
    global: {
      stubs: {
        UiButton,
        UiDialog,
      },
    },
  });
});

describe('ui_confirm_dialog', () => {
  it('should render correctly', () => {
    expect(wrapper.findComponent(UiConfirmDialog)).toBeTruthy();
  });

  it('should render correctly with content', async () => {
    await wrapper.trigger();

    expect(wrapper.find('[data-test="ui-confirm-dialog-header"]'));
  });
});
