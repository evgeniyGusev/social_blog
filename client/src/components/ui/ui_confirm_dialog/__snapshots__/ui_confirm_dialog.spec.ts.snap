// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ui_confirm_dialog > correct render props 1`] = `"<teleport-stub data-v-d40695e1="" data-v-3297a1c9="" to="body" class="ui-confirm-dialog" closable="false"></teleport-stub>"`;

exports[`ui_confirm_dialog > should render correctly 1`] = `"<teleport-stub data-v-d40695e1="" data-v-3297a1c9="" to="body" class="ui-confirm-dialog" closable="false"></teleport-stub>"`;
